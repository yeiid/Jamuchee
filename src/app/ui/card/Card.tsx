"use client"
import  { useState } from 'react';
import Cart from './Cart';
import { useCart } from '@/context/Contex';



const Card: React.FC = () => {
  const [isCartOpen, setCartOpen] = useState(false);
  const {cart} = useCart()
  const productCount = cart.reduce((total, product) => total + product.quantity, 0);

  const handleToggleCart = () => {
    setCartOpen(!isCartOpen);
  };
  return (
<>
      <div className=" flex h-[48px] grow items-center justify-center gap-2 rounded-md bg-gray-50 p-3 text-sm font-medium hover:bg-sky-100 hover:text-blue-600 md:flex-none md:justify-start md:p-2 md:px-3">
        <button className="mt-2 bg-gray-700 text-white px-4 py-2 rounded" onClick={handleToggleCart}>ğŸ›’cart {productCount}</button>
      </div>
      <Cart isOpen={isCartOpen} onClose={() => setCartOpen(false)} />
</>  );
};

export default Card;
