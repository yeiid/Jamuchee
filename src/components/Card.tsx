"use client";
import React, { useState } from 'react';
import { useCart } from "@/context/Contex";
import { ShoppingCart, Plus, Minus, Trash2, Link } from 'lucide-react';

const CartComponent: React.FC = () => {
  const [isCartOpen, setCartOpen] = useState(false);
  const { cart, dispatch } = useCart();

  const handleToggleCart = () => {
    setCartOpen(!isCartOpen);
  };

  const handleAddToCart = (productId: string) => {
    dispatch({
      type: "ADD_TO_CART",
      payload: { id: productId },
    });
  };

  const handleDecreaseQuantity = (productId: string) => {
    dispatch({
      type: "DECREASE_QUANTITY",
      payload: productId,
    });
  };

  const handleRemoveFromCart = (productId: string) => {
    dispatch({
      type: "REMOVE_FROM_CART",
      payload: productId,
    });
  };

  const productCount = cart.reduce(
    (total, product) => total + product.quantity,
    0
  );

  const getTotalPrice = () => {
    return cart.reduce(
      (total, product) => total + product.quantity * product.price,
      0
    );
  };

  const handleBuy = () => {
    if (cart.length === 0) {
      alert("El carrito está vacío");
      return;
    }
  
    const cartDetails = cart.map(
      (item) => `${item.quantity} x ${item.name} - $${item.price * item.quantity}`
    );
    const message = `¡Hola! Estoy interesado en comprar:\n${cartDetails.join(
      "\n"
    )}\nTotal: $${getTotalPrice()}`;
  
    const phoneNumber = "3160457939";
    const whatsappLink = `https://wa.me/${phoneNumber}?text=${encodeURIComponent(
      message
    )}`;
  
    window.open(whatsappLink, '_blank');
  };

  return (
    <>
      <button
        className="flex items-center space-x-1 bg-green-600 text-white px-4 py-2 rounded-full hover:bg-green-700 transition duration-300"
        onClick={handleToggleCart}
      >
        <ShoppingCart size={20} />
        <span>{productCount}</span>
      </button>

      {isCartOpen && (
        <div className="fixed inset-0 bg-black bg-opacity-50 z-50">
          <div className="fixed top-0 right-0 h-full w-full max-w-md bg-white shadow-lg transform transition-transform translate-x-0 overflow-y-auto">
            <div className="p-4 bg-green-600 text-white">
              <h2 className="text-2xl font-bold">Carrito de Compras</h2>
              <button
                className="absolute top-4 right-4 text-xl cursor-pointer"
                onClick={handleToggleCart}
              >
                &times;
              </button>
            </div>
            <div className="p-4">
              {cart.length === 0 ? (
                <p className="text-center text-gray-500 my-4">El carrito está vacío</p>
              ) : (
                <>
                  {cart.map((product) => (
                    <div key={product.id} className="flex justify-between items-center border-b py-4">
                      <div>
                        <p className="font-semibold">{product.name}</p>
                        <p className="text-sm text-gray-500">${product.price} x {product.quantity}</p>
                      </div>
                      <div className="flex items-center space-x-2">
                        <button
                          className="p-1 rounded-full bg-green-100 text-green-600 hover:bg-green-200"
                          onClick={() => handleDecreaseQuantity(product.id)}
                        >
                          <Minus size={16} />
                        </button>
                        <span>{product.quantity}</span>
                        <button
                          className="p-1 rounded-full bg-green-100 text-green-600 hover:bg-green-200"
                          onClick={() => handleAddToCart(product.id)}
                        >
                          <Plus size={16} />
                        </button>
                        <button
                          className="p-1 rounded-full bg-red-100 text-red-600 hover:bg-red-200"
                          onClick={() => handleRemoveFromCart(product.id)}
                        >
                          <Trash2 size={16} />
                        </button>
                      </div>
                    </div>
                  ))}
                  <div className="mt-4">
                    <p className="text-xl font-bold">Total: ${getTotalPrice()}</p>
                    <button
                      className="mt-4 w-full bg-green-600 text-white font-bold py-2 px-4 rounded hover:bg-green-700 transition duration-300"
                      >
                    
                    </button>
                    <Link href='/src/app/ui/products/productDetail/page.tsx'/>
                  </div>
                </>
              )}
            </div>
          </div>
        </div>
      )}
    </>
  );
};

export default CartComponent;