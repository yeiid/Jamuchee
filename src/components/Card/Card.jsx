import React from "react";

const Card = ({ title, subtitle, imageUrl, children }) => {
  return (
    <div className="card">
      <div className="card-header">
        <h2 className="card-title">{title}</h2>
        {subtitle && <h3 className="card-subtitle">{subtitle}</h3>}
      </div>

      {imageUrl && (
        <div className="card-image">
          <img src={imageUrl} alt={title} />
        </div>
      )}

      {children && <div className="card-content">{children}</div>}
    </div>
  );
};

export default Card;
