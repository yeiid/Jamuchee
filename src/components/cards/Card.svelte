<script>
  /**
   * @typedef {Object} Props
   * @property {string} title - Título de la tarjeta
   * @property {string} [subtitle] - Subtítulo opcional de la tarjeta
   * @property {string} [imageUrl] - URL de la imagen opcional
   */

  /** @type {string} */
  export let title = '';
  
  /** @type {string | undefined} */
  export let subtitle = undefined;
  
  /** @type {string | undefined} */
  export let imageUrl = undefined;
</script>

<div class="card p-4">
  <div class="card-header mb-4">
    <h2 class="card-title text-xl font-bold">{title}</h2>
    {#if subtitle}
      <h3 class="card-subtitle text-gray-600 dark:text-gray-400">{subtitle}</h3>
    {/if}
  </div>
  
  {#if imageUrl}
    <div class="card-image mb-4">
      <img src={imageUrl} alt={title} class="w-full rounded-md" />
    </div>
  {/if}
  
  <div class="card-content">
    <slot />
  </div>
</div> 