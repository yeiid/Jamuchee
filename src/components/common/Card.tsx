"use client";
import React, { useState, useEffect, useCallback } from "react";
import { ShoppingCart, X, ShoppingBag } from "lucide-react";
import Link from "next/link";
import CartItem from "./CartItem";
import useCartStore, { CartItem as CartItemType } from "@/store/useCartStore";
import { useRouter } from "next/navigation";

const CartComponent: React.FC = () => {
  const router = useRouter();
  const [isCartOpen, setCartOpen] = useState(false);
  const [isMounted, setIsMounted] = useState(false);

  // Usar directamente el store de Zustand en lugar del contexto
  const items = useCartStore((state) => state.items);
  const addToCart = useCartStore((state) => state.addToCart);
  const removeFromCart = useCartStore((state) => state.removeFromCart);
  const updateQuantity = useCartStore((state) => state.updateQuantity);
  const getTotalItems = useCartStore((state) => state.getTotalItems);
  const getTotalPrice = useCartStore((state) => state.getTotalPrice);

  // Este useEffect asegura que el componente solo se renderice en el cliente
  useEffect(() => {
    setIsMounted(true);
  }, []);

  // Evita el scroll cuando el carrito está abierto
  useEffect(() => {
    document.body.style.overflow = isCartOpen ? "hidden" : "unset";
    return () => {
      document.body.style.overflow = "unset";
    };
  }, [isCartOpen]);

  const handleToggleCart = useCallback(() => setCartOpen((prev) => !prev), []);

  const handleAddToCart = useCallback(
    (productId: string) => {
      const item = items.find((item) => item.id === productId);
      if (item) {
        updateQuantity(productId, item.quantity + 1);
      }
    },
    [items, updateQuantity]
  );

  const handleDecreaseQuantity = useCallback(
    (productId: string) => {
      const item = items.find((item) => item.id === productId);
      if (item && item.quantity > 1) {
        updateQuantity(productId, item.quantity - 1);
      } else if (item) {
        removeFromCart(productId);
      }
    },
    [items, updateQuantity, removeFromCart]
  );

  const handleRemoveFromCart = useCallback(
    (productId: string) => {
      removeFromCart(productId);
    },
    [removeFromCart]
  );

  // Redireccionar a la página de checkout
  const goToCheckout = useCallback(() => {
    setCartOpen(false);
    router.push("/checkout");
  }, [router]);

  // Si el componente no está montado, renderiza un placeholder para evitar errores de hidratación
  if (!isMounted) {
    return (
      <button
        className="flex items-center space-x-2 bg-primary-600 text-white px-4 py-2 rounded-full hover:bg-primary-700 transition-colors duration-300 shadow-lg"
        aria-label="Ver carrito"
      >
        <ShoppingCart size={20} />
      </button>
    );
  }

  const productCount = getTotalItems();
  const totalPrice = getTotalPrice().toFixed(2);

  return (
    <>
      <button
        className="flex items-center space-x-2 bg-primary-600 text-white px-4 py-2 rounded-full hover:bg-primary-700 transition-all duration-300 shadow-lg hover:shadow-xl hover:scale-105"
        onClick={handleToggleCart}
        aria-label="Ver carrito"
      >
        <ShoppingCart size={20} />
        {productCount > 0 && (
          <div className="bg-white dark:bg-gray-800 text-primary-600 dark:text-primary-400 rounded-full w-5 h-5 flex items-center justify-center text-xs font-bold ml-2 animate-bounce">
            {productCount}
          </div>
        )}
      </button>

      {/* Overlay del carrito */}
      <div
        className={`fixed inset-0 z-50 ${
          isCartOpen
            ? "opacity-100 pointer-events-auto"
            : "opacity-0 pointer-events-none"
        } bg-black bg-opacity-50 transition-all duration-300 backdrop-blur-sm`}
        onClick={handleToggleCart}
      >
        {/* Contenido del carrito */}
        <div
          className={`fixed top-0 right-0 h-full max-w-md w-full bg-white dark:bg-gray-900 shadow-2xl transform transition-all duration-300 ${
            isCartOpen ? "translate-x-0" : "translate-x-full"
          } flex flex-col`}
          onClick={(e) => e.stopPropagation()}
        >
          {/* Header */}
          <div className="p-4 bg-gradient-to-r from-primary-600 to-primary-700 text-white flex justify-between items-center">
            <h2 className="text-2xl font-bold">Carrito de Compras</h2>
            <button
              onClick={handleToggleCart}
              className="hover:text-primary-200 transition-colors duration-200 p-2 rounded-full hover:bg-primary-500"
              aria-label="Cerrar carrito"
            >
              <X size={24} />
            </button>
          </div>

          {/* Lista de productos */}
          <div className="flex-grow overflow-y-auto p-4">
            {items.length === 0 ? (
              <div className="flex flex-col items-center justify-center h-full text-gray-500 dark:text-gray-400">
                <ShoppingBag
                  size={64}
                  className="mb-4 text-primary-600 dark:text-primary-400 opacity-50 animate-pulse"
                />
                <p className="text-center text-lg font-medium">
                  Tu carrito está vacío
                </p>
                <p className="text-center text-sm mt-2">
                  ¡Agrega algunos productos maravillosos!
                </p>
                <Link href="/features/products" className="mt-4">
                  <button className="px-6 py-2 bg-primary-600 text-white rounded-full hover:bg-primary-700 transition-all duration-300 hover:shadow-lg hover:scale-105">
                    Ver productos
                  </button>
                </Link>
              </div>
            ) : (
              items.map((product: CartItemType) => (
                <CartItem
                  key={product.id}
                  product={product}
                  handleAddToCart={handleAddToCart}
                  handleDecreaseQuantity={handleDecreaseQuantity}
                  handleRemoveFromCart={handleRemoveFromCart}
                />
              ))
            )}
          </div>

          {/* Footer */}
          <div className="p-4 bg-gradient-to-b from-white to-primary-50 dark:from-gray-900 dark:to-gray-800 border-t border-primary-200 dark:border-gray-700">
            <div className="flex justify-between mb-2">
              <span className="text-gray-600 dark:text-gray-300">
                Subtotal:
              </span>
              <span className="font-medium text-gray-900 dark:text-white">
                ${totalPrice}
              </span>
            </div>
            <div className="flex justify-between mb-4">
              <span className="text-gray-600 dark:text-gray-300">
                Productos:
              </span>
              <span className="font-medium text-gray-900 dark:text-white">
                {productCount}
              </span>
            </div>
            <hr className="border-primary-200 dark:border-gray-700 mb-4" />
            <p className="text-xl font-bold text-primary-800 dark:text-primary-400 mb-4">
              Total: ${totalPrice}
            </p>
            <button
              onClick={goToCheckout}
              className="w-full bg-primary-600 text-white font-bold py-3 px-4 rounded-lg hover:bg-primary-700 transition-all duration-300 shadow-lg hover:shadow-xl hover:scale-105 disabled:opacity-50 disabled:cursor-not-allowed disabled:hover:scale-100 disabled:hover:shadow-lg flex items-center justify-center"
              disabled={items.length === 0}
            >
              <ShoppingBag size={20} className="mr-2" />
              Finalizar Pedido
            </button>
          </div>
        </div>
      </div>
    </>
  );
};

export default CartComponent;
