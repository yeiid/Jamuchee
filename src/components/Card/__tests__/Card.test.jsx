import React from "react";
import { render } from "@testing-library/react";
import Card from "../Card";

describe("Componente Card", () => {
  test("renderiza correctamente con props básicas", () => {
    const { container } = render(
      <Card
        title="Título de tarjeta"
        subtitle="Subtítulo opcional"
        imageUrl="/path/to/image.jpg"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("renderiza correctamente sin imagen", () => {
    const { container } = render(<Card title="Solo título" />);
    expect(container).toMatchSnapshot();
  });

  test("renderiza contenido hijo correctamente", () => {
    const { container } = render(
      <Card title="Tarjeta con hijos">
        <p>Este es un párrafo dentro de la tarjeta</p>
        <button>Un botón</button>
      </Card>
    );
    expect(container).toMatchSnapshot();
  });
});
