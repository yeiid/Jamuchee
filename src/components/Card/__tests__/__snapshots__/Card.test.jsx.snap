// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente Card renderiza contenido hijo correctamente 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <h2
        class="card-title"
      >
        Tarjeta con hijos
      </h2>
    </div>
    <div
      class="card-content"
    >
      <p>
        Este es un párrafo dentro de la tarjeta
      </p>
      <button>
        Un botón
      </button>
    </div>
  </div>
</div>
`;

exports[`Componente Card renderiza correctamente con props básicas 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <h2
        class="card-title"
      >
        Título de tarjeta
      </h2>
      <h3
        class="card-subtitle"
      >
        Subtítulo opcional
      </h3>
    </div>
    <div
      class="card-image"
    >
      <img
        alt="Título de tarjeta"
        src="/path/to/image.jpg"
      />
    </div>
  </div>
</div>
`;

exports[`Componente Card renderiza correctamente sin imagen 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <h2
        class="card-title"
      >
        Solo título
      </h2>
    </div>
  </div>
</div>
`;
